src/components/Tooltip.tsx

```diff
@@ -0,0 +1,121 @@
+import React, { ReactNode, CSSProperties } from 'react'
+import Portal from './Portal'
+import { usePopperDefault, TOOLTIP_OFFSET } from 'hooks/usePopper'
+import { State, Placement } from '@popperjs/core'
+import styled from 'styled-components'
+import { isElement, isFragment } from 'react-is'
+
+// visibility necessary for correct boundingRect calculation by popper
+const TooltipOuter = styled.div<Pick<TooltipBaseProps, 'isShown'>>`
+  visibility: ${(props): 'hidden' | false => !props.isShown && 'hidden'};
+`
+// can style anything but TOOLTIP_OFFSET fields, position and transform: rotate
+const TooltipArrow = styled.div`
+  &,
+  ::before {
+    position: absolute;
+    width: ${TOOLTIP_OFFSET}px;
+    height: ${TOOLTIP_OFFSET}px;
+    z-index: -1;
+  }
+
+  ::before {
+    content: '';
+    transform: rotate(45deg);
+    background: #333;
+  }
+`
+
+const TooltipInner = styled.div`
+  background: #333;
+  color: white;
+  font-weight: bold;
+  padding: 4px 8px;
+  font-size: 13px;
+  border-radius: 4px;
+
+  &[data-popper-placement^='top'] > ${TooltipArrow} {
+    bottom: -${TOOLTIP_OFFSET / 2}px;
+  }
+
+  &[data-popper-placement^='bottom'] > ${TooltipArrow} {
+    top: -${TOOLTIP_OFFSET / 2}px;
+  }
+
+  &[data-popper-placement^='left'] > ${TooltipArrow} {
+    right: -${TOOLTIP_OFFSET / 2}px;
+  }
+
+  &[data-popper-placement^='right'] > ${TooltipArrow} {
+    left: -${TOOLTIP_OFFSET / 2}px;
+  }
+`
+
+interface TooltipBaseProps {
+  isShown: boolean
+  state: Partial<Pick<State, 'placement' | 'styles'>>
+}
+
+const TooltipBase: React.FC<TooltipBaseProps> = ({ children, isShown, state }, ref) => {
+  const { placement, styles = {} } = state
+
+  return (
+    // Portal isolates popup styles from the App styles
+    <Portal>
+      <TooltipOuter isShown={isShown}>
+        <TooltipInner role="tooltip" ref={ref} style={styles.popper as CSSProperties} data-popper-placement={placement}>
+          <TooltipArrow data-popper-arrow style={styles.arrow as CSSProperties} />
+          {isShown && children}
+        </TooltipInner>
+      </TooltipOuter>
+    </Portal>
+  )
+}
+
+interface TooltipProps extends TooltipBaseProps {
+  children?: ReactNode
+}
+
+export const Tooltip = React.memo(React.forwardRef<HTMLDivElement, TooltipProps>(TooltipBase))
+
+interface WrapperProps {
+  tooltip: ReactNode
+  placement?: Placement
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  as?: keyof JSX.IntrinsicElements | React.ComponentType<any>
+}
+
+const Wrapper: React.FC<WrapperProps> = ({ children, tooltip, placement, as }) => {
+  const { targetProps, tooltipProps } = usePopperDefault<HTMLDivElement>(placement)
+
+  const chilrenNumber = React.Children.count(children)
+
+  // can attach ref to element
+  // if children is a single element, not just text, and `as` tag|Component not specified
+  if (chilrenNumber === 1 && typeof as === 'undefined' && isElement(children) && !isFragment(children)) {
```

Comment:
what is this for? seems like crazy amount of adding of code

Reply:
To be able to clone Element later and pass it a `ref`


> not saying it's a bad thing, but there's a lot seemingly going on in this PR that seems to be adding to the library. The hook looks very complicated - what was the reasoning behind adding it vs using the library directly?

Library directly do anything with React.